src/txmempool.cpp

```diff
@@ -784,33 +784,101 @@ void CTxMemPool::_removeConflicts(const CTransaction &tx, std::list<CTransaction
+// data in mapLinks and only removing transactions with no ancestors.
+// If a transaction has an ancestor, it is placed on a deferred map: ancestor->descendant list.
+// Whenever a tx is removed from the mempool, any of its descendants are placed back onto the
+// queue of tx that are removable.
 void CTxMemPool::removeForBlock(const std::vector<CTransactionRef> &vtx,
     unsigned int nBlockHeight,
     std::list<CTransactionRef> &conflicts,
     bool fCurrentEstimate)
 {
+    typedef std::map<txiter, std::vector<txiter>, CompareIteratorByHash> DeferredMap;
+
     WRITELOCK(cs);
+    std::vector<txiter> entriesToRemove;
     std::vector<CTxMemPoolEntry> entries;
+
+    DeferredMap deferred;
+    std::queue<txiter> ready;
+
     for (const auto &tx : vtx)
     {
+        setEntries descendants;
         uint256 hash = tx->GetHash();
-
-        indexed_transaction_set::iterator i = mapTx.find(hash);
-        if (i != mapTx.end())
-            entries.push_back(*i);
+        txiter i = mapTx.find(hash);
+        if (i == mapTx.end())
+            continue; // not in the mempool
+        ready.push(i);
     }
-    for (const auto &tx : vtx)
+
+    while (!ready.empty())
     {
-        txiter it = mapTx.find(tx->GetHash());
-        if (it != mapTx.end())
+        txiter nextTx = ready.front();
+        ready.pop();
+
+        const auto &links = mapLinks.find(nextTx);
+        if (links == mapLinks.end())
         {
+            DbgAssert(!"should never happen", );
+            continue;
+        }
+        if (links->second.parents.empty()) // This tx has no ancestors, so its ready to be removed
+        {
+            entries.push_back(*nextTx); // save tx for estimator adjustment at the bottom
+
+            // place any tx that were deferred because this tx was required back onto the ready Q
+            DeferredMap::iterator deferredSet = deferred.find(nextTx);
+            if (deferredSet != deferred.end())
+            {
+                for (auto j = deferredSet->second.begin(); j != deferredSet->second.end(); ++j)
+                {
+                    ready.push(*j);
+                }
+                deferred.erase(deferredSet);
+            }
+            // Remove this tx from the mempool
             setEntries stage;
-            stage.insert(it);
+            stage.insert(nextTx); // nextTx is invalid now
             _RemoveStaged(stage, true);
         }
+        else // This tx has ancestors, so put it in the deferred queue waiting for them
+        {
+            auto tmp = links->second.parents.begin();
+            if (tmp != links->second.parents.end())
+            {
+                txiter t = *tmp;
+                std::vector<txiter> &vec = deferred[t];
+                vec.push_back(nextTx);
+            }
+            else // an empty txiter should never be part of the parents
+            {
+                DbgAssert("tx parents is corrupt!", ready.push(nextTx));
+            }
```

Comment:
I wonder whether this dbgassert could cause an infinite loop with the ready.push(nextTx) ??

Reply:
good point.  I will delete clear the parents since the data structure is suspect anyway.